File: Leetcode/hashing/counting/missing-number/missingNumber.ts
function missingNumber(nums: number[]): number {
    const map = new Map();

    for (let i = 0; i <= nums.length; i++) {
        map.set(nums[i], nums[i]);
    }

    for (let i = 0; i <= nums.length; i++) {
        if (!map.has(i)) return i;
    }

    return -1;
}

// Time: O(n);
// Space: O(n);

const result = missingNumber([3, 0, 1]);
console.log(result);

File: Leetcode/hashing/existing/missing-number/missingNumber.java
public int missingNumber(int[] nums) {
    Set<Integer> numsSet = new HashSet<>();

    for (int num: nums) {
        numsSet.add(num);
    }

    for (int i = 0; i < nums.length + 1; i++) {
        if (!numsSet.contains(i)) {
            return i;
        }
    }

    return -1;
}

//Time complexity : O(n)

//Because the set allows for O(1) containment queries, the main loop
//runs in O(n) time. Creating num_set costs O(n) time, as each set insertion
//runs in amortized O(1) time, so the overall runtime is O(n+n)=O(n).

//Space complexity : O(n)
